File: testdata/fixtures/swift/CWE-78.swift
// CWE-78: the shell gets a string somebody else wrote.
//
// This is macOS Swift rather than iOS - a sandboxed app cannot spawn a process
// at all - and it is the same weakness it is everywhere. `Process` with
// `/bin/sh -c` hands the whole argument to a shell, so a `;` or a backtick in
// the middle of it starts a second command with the app's privileges.
//
// Passing the program and its arguments separately is the fix, and it is the
// default: `executableURL` plus an `arguments` array never involves a shell,
// so a semicolon in an argument is just a semicolon.
import Foundation

final class Tools {

    func convertFromDeepLink(_ url: URL) {
        let name = url.lastPathComponent
        let task = Process()
        task.executableURL = URL(fileURLWithPath: "/bin/sh")
        // ruleid: swift-command-from-untrusted-input
        task.arguments = ["-c", "convert \(name) out.png"]
        try? task.run()
    }

    func inspectFromPasteboard() {
        let target = UIPasteboard.general.string ?? ""
        let task = Process()
        task.launchPath = "/bin/bash"
        // ruleid: swift-command-from-untrusted-input
        task.arguments = ["-c", "file " + target]
        task.launch()
    }

    func convertFixed() {
        let task = Process()
        task.executableURL = URL(fileURLWithPath: "/bin/sh")
        // ok: swift-command-from-untrusted-input
        task.arguments = ["-c", "convert in.png out.png"]
        try? task.run()
    }

    func convertSeparately(_ url: URL) {
        let name = url.lastPathComponent
        let task = Process()
        task.executableURL = URL(fileURLWithPath: "/usr/bin/convert")
        // ok: swift-command-from-untrusted-input
        task.arguments = [name, "out.png"]
        try? task.run()
    }
}

File: testdata/fixtures/python/CWE-78.py
# CWE-78: OS command injection — the shell gets a string the caller chose.
#
# `os.system`, `os.popen`, `subprocess` with `shell=True`, and `eval` of a
# shell command all hand the whole command line to `/bin/sh`. A `;`, a
# backtick or a `$(...)` in the middle starts a second command.
# Passing the program and its arguments separately — `subprocess.run([prog, arg])`
# with `shell=False` — never involves a shell, so a semicolon is just a character.
import os
import subprocess
from flask import request


@app.route("/run/<cmd>")
def from_route(cmd):
    # ruleid: python-os-command-injection
    os.system(cmd)
    # ruleid: python-os-command-injection
    subprocess.call(cmd, shell=True)
    # ruleid: python-os-command-injection
    subprocess.Popen("echo " + cmd, shell=True)


def from_query():
    q = request.args.get("q")
    # ruleid: python-os-command-injection
    os.popen(q)
    # ruleid: python-os-command-injection
    subprocess.run(q, shell=True)
    # Propagation through a variable — reaching must follow it.
    cmd = q
    # ruleid: python-os-command-injection
    os.system(cmd)


def from_form():
    data = request.form.get("data")
    tainted = data
    # ruleid: python-os-command-injection
    subprocess.call(tainted, shell=True)


def safe_literal():
    # ok: python-os-command-injection
    os.system("ls -l")
    # ok: python-os-command-injection
    subprocess.call(["ls", "-l"])
    # ok: python-os-command-injection
    subprocess.run(["echo", "hello"], shell=False)
    # ok: python-os-command-injection
    subprocess.Popen(["/bin/ls", "-l"])


def safe_constant():
    cmd = "uptime"
    # ok: python-os-command-injection
    os.system("uptime | head")


# `input()` is the other way a caller chooses a command, and it is the same
# weakness: a prompt is untrusted for exactly the reason a query string is.
def from_a_prompt():
    cmd = input()
    # ruleid: python-os-command-injection
    os.system(cmd)


def from_a_prompt_concatenated():
    data = input("enter: ")
    # ruleid: python-os-command-injection
    os.system(data + "; echo hi")


def from_a_literal():
    x = "fixed"
    # ok: python-os-command-injection
    os.system(x)
